File: app/web/src/api/sdf/dal/action.ts
import { FuncRunId } from "@/newhotness/api_composables/func_run";
import { ChangeSetId } from "./change_set";
import { ComponentId } from "./component";

export type ActionPrototypeId = string;
export type ActionId = string;

export enum ActionState {
  Dispatched = "Dispatched",
  Failed = "Failed",
  OnHold = "OnHold",
  Queued = "Queued",
  Running = "Running",
}

export enum ActionKind {
  Create = "Create",
  Destroy = "Destroy",
  Refresh = "Refresh",
  Manual = "Manual",
  Update = "Update",
}

export interface ActionPrototype {
  id: ActionPrototypeId;
  name: string;
  displayName: string;
}

export type ActionResultState = "Success" | "Failure" | "Unknown";

export interface ActionView {
  id: ActionId;
  actor?: string; // TODO i dont see this on the backend
  prototypeId: ActionPrototypeId;
  componentId: ComponentId | null;
  name: string;
  description?: string;
  kind: ActionKind;
  originatingChangeSetId: ChangeSetId;
  funcRunId?: FuncRunId;
}

export interface ActionProposedView extends ActionView {
  state: ActionState;
  myDependencies: ActionId[];
  dependentOn: ActionId[];
  holdStatusInfluencedBy: ActionId[];
  componentSchemaName?: string;
  componentName?: string;
}

File: lib/si-events-rs/src/action.rs
use serde::{
    Deserialize,
    Serialize,
};
pub use si_id::{
    ActionId,
    ActionPrototypeId,
};
use strum::{
    AsRefStr,
    Display,
    EnumDiscriminants,
};

#[remain::sorted]
#[derive(AsRefStr, Debug, Copy, Clone, Deserialize, Serialize, PartialEq, Eq, Display, Hash)]
pub enum ActionKind {
    /// Create the "outside world" version of the modeled object.
    Create,
    /// Destroy the "outside world" version of the modeled object referenced in the resource.
    Destroy,
    /// This [`Action`][crate::Action] will only ever be manually queued.
    Manual,
    /// Refresh the resource to reflect the current state of the modeled object in the "outside
    /// world".
    Refresh,
    /// Update the version of the modeled object in the "outside world" to match the state of the
    /// model.
    Update,
}

#[derive(
    Debug,
    Copy,
    Clone,
    Deserialize,
    Serialize,
    EnumDiscriminants,
    PartialEq,
    Eq,
    Display,
    Hash,
    Default,
)]
#[strum_discriminants(derive(strum::Display, Serialize, Deserialize))]
pub enum ActionState {
    /// Action has been determined to be eligible to run, and has had its job sent to the job
    /// queue.
    Dispatched,
    /// Action failed during execution. See the job history for details.
    Failed,
    /// Action is "queued", but should not be considered as eligible to run, until moved to the
    /// `Queued` state.
    OnHold,
    /// Action is available to be dispatched once all of its prerequisites have succeeded, and been
    /// removed from the graph.
    #[default]
    Queued,
    /// Action has been dispatched, and started execution in the job system. See the job history
    /// for details.
    Running,
}

#[derive(Debug, Copy, Clone, PartialEq, Eq, Display, Serialize, Deserialize)]
pub enum ActionResultState {
    Success,
    Failure,
    Unknown,
}
